### PR TITLE
add document.implementation

### PR DESCRIPTION
**Pull request by krichprollsch**
Relates to https://github.com/Browsercore/browsercore/issues/18

Depends on https://github.com/Browsercore/jsruntime-lib/pull/146

### CODE REVIEW

File: src/netsurf.zig [hunk 1]
```diff
@@ -768,6 +768,54 @@ pub inline fn documentTypeGetSystemId(dt: *DocumentType) []const u8 {
...
 }
 
+// DOMImplementation
+pub inline fn domImplementationCreateDocument(namespace: ?[:0]const u8, qname: ?[:0]const u8, doctype: ?*DocumentType) *Document {
+    var doc: ?*Document = undefined;
+
+    var ptrnamespace: [*c]const u8 = null;
+    if (namespace != null) {
+        ptrnamespace = namespace.?.ptr;
+    }
```
Comment (francisbouvier) [inline src/netsurf.zig]:
```suggestion
    if (namespace) |ns| {
        ptrnamespace = ns.ptr;
    }
```

File: src/netsurf.zig [hunk 2]
```diff
@@ -768,6 +768,54 @@ pub inline fn documentTypeGetSystemId(dt: *DocumentType) []const u8 {
...
+
+    var ptrqname: [*c]const u8 = null;
+    if (qname != null) {
+        ptrqname = qname.?.ptr;
+    }
```
Comment (francisbouvier) [inline src/netsurf.zig]:
```suggestion
    var ptrqname: [*c]const u8 = null;
    if (qname) |qn| {
        ptrqname = qn.ptr;
    }
```

File: src/netsurf.zig [hunk 3]
```diff
@@ -768,6 +768,54 @@ pub inline fn documentTypeGetSystemId(dt: *DocumentType) []const u8 {
...
+}
+
+pub inline fn domImplementationCreateDocumentType(qname: [:0]const u8, publicId: [:0]const u8, systemId: [:0]const u8) *DocumentType {
+    var dt: ?*DocumentType = undefined;
+    _ = c.dom_implementation_create_document_type(qname.ptr, publicId.ptr, systemId.ptr, &dt);
+    return dt.?;
+}
+
+pub inline fn domImplementationCreateHTMLDocument(title: ?[]const u8) *Document {
+    _ = title;
```
Comment (francisbouvier) [inline src/netsurf.zig]:
```suggestion
```

File: src/netsurf.zig [hunk 4]
```diff
@@ -768,6 +768,54 @@ pub inline fn documentTypeGetSystemId(dt: *DocumentType) []const u8 {
...
+    _ = c.dom_implementation_create_document(
+        c.DOM_IMPLEMENTATION_HTML,
+        null,
+        null,
+        null,
+        null,
+        null,
+        &doc,
+    );
+    // TODO set title
```
Comment (francisbouvier) [inline src/netsurf.zig:872]:
```suggestion
    // TODO set title
    _ = title;
```

File: src/dom/implementation.zig [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+const Document = @import("document.zig").Document;
+const DocumentType = @import("document_type.zig").DocumentType;
+
+// WEB IDL https://dom.spec.whatwg.org/#domimplementation
+pub const DOMImplementation = struct {
+    pub const mem_guarantied = true;
+
+    pub fn _createDocumentType(
+        self: *DOMImplementation,
+        allocator: std.mem.Allocator,
```
Comment (francisbouvier) [inline src/dom/implementation.zig]:
```suggestion
        alloc: std.mem.Allocator,
```

Everywhere else we use the short version `alloc`

File: src/dom/implementation.zig [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        defer allocator.free(cqname);
+
+        const cpublicId = try allocator.dupeZ(u8, publicId);
+        defer allocator.free(cpublicId);
+
+        const csystemId = try allocator.dupeZ(u8, systemId);
+        defer allocator.free(csystemId);
+
+        const dt = parser.domImplementationCreateDocumentType(cqname, cpublicId, csystemId);
+        return dt;
```
Comment (francisbouvier) [inline src/dom/implementation.zig]:
```suggestion
        return parser.domImplementationCreateDocumentType(cqname, cpublicId, csystemId);
```

File: src/dom/implementation.zig [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    pub fn _createDocument(
+        self: *DOMImplementation,
+        allocator: std.mem.Allocator,
```
Comment (francisbouvier) [inline src/dom/implementation.zig]:
```suggestion
        alloc: std.mem.Allocator,
```

File: src/dom/implementation.zig [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+const Document = @import("document.zig").Document;
+const DocumentType = @import("document_type.zig").DocumentType;
+
+// WEB IDL https://dom.spec.whatwg.org/#domimplementation
+pub const DOMImplementation = struct {
+    pub const mem_guarantied = true;
+
+    pub fn _createDocumentType(
+        self: *DOMImplementation,
```
Comment (francisbouvier) [inline src/dom/implementation.zig]:
```suggestion
       _: *DOMImplementation,
```

If it's a permanent state not to use `self` I think it's better to put the `_` in the function definition than in the function body.

File: src/dom/implementation.zig [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+        const csystemId = try allocator.dupeZ(u8, systemId);
+        defer allocator.free(csystemId);
+
+        const dt = parser.domImplementationCreateDocumentType(cqname, cpublicId, csystemId);
+        return dt;
+    }
+
+    pub fn _createDocument(
+        self: *DOMImplementation,
```
Comment (francisbouvier) [inline src/dom/implementation.zig]:
```suggestion
       _: *DOMImplementation,
```

File: src/dom/implementation.zig [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        namespace: ?[]const u8,
+        qname: ?[]const u8,
+        doctype: ?*parser.DocumentType,
+    ) !*parser.Document {
+        _ = self;
+        var cnamespace: ?[:0]const u8 = null;
+        if (namespace != null) {
+            cnamespace = try allocator.dupeZ(u8, namespace.?);
+            defer allocator.free(cnamespace.?);
+        }
```
Comment (francisbouvier) [inline src/dom/implementation.zig]:
```suggestion
        if (namespace) |ns| {
            cnamespace = try allocator.dupeZ(u8, ns);
            defer allocator.free(cnamespace.?);
        }
```

File: src/dom/implementation.zig [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+        var cqname: ?[:0]const u8 = null;
+        if (qname != null) {
+            cqname = try allocator.dupeZ(u8, qname.?);
+            defer allocator.free(cqname.?);
+        }
```
Comment (francisbouvier) [inline src/dom/implementation.zig]:
```suggestion
        if (qname) |qn| {
            cqname = try allocator.dupeZ(u8, qn);
            defer allocator.free(cqname.?);
        }
```

File: src/dom/implementation.zig [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+        const doc = parser.domImplementationCreateDocument(cnamespace, cqname, doctype);
+        return doc;
```
Comment (francisbouvier) [inline src/dom/implementation.zig]:
```suggestion
        return parser.domImplementationCreateDocument(cnamespace, cqname, doctype);
```

File: src/dom/implementation.zig [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    pub fn _createHTMLDocument(_: *DOMImplementation, title: ?[]const u8) *parser.Document {
+        const doc = parser.domImplementationCreateHTMLDocument(title);
+        return doc;
```
Comment (francisbouvier) [inline src/dom/implementation.zig]:
```suggestion
        return parser.domImplementationCreateHTMLDocument(title);
```

**Comment by krichprollsch** [comment]
@francisbouvier PR updated and rebased

File: src/dom/implementation.zig [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+    pub fn _createDocumentType(
+        _: *DOMImplementation,
+        alloc: std.mem.Allocator,
+        qname: []const u8,
+        publicId: []const u8,
+        systemId: []const u8,
+    ) !*parser.DocumentType {
+        const cqname = try alloc.dupeZ(u8, qname);
+        defer alloc.free(cqname);
```
Comment (francisbouvier) [inline src/dom/implementation.zig:24]: Do we need to alloc here if we free after function returns?

Comment (krichprollsch) [reply src/dom/implementation.zig:24]: As discussed via Slack, I use the allocator to create a null ending string from the slice.

**Pull request merged**